ruby/vending-machine/lib/vending_machine.rb

```diff
@@ -0,0 +1,133 @@
+require_relative 'vending_machine/product_tray'
+require_relative 'vending_machine/coin_sorter'
+
+class VendingMachine
+  InvalidCodeError = Class.new(StandardError)
+  PaymentRequiredError = Class.new(StandardError)
+
+  attr_reader :product_trays, :coin_sorter, :display
+
+  INIT_PRICES = [80, 120, 160].freeze
+  NO_PRODUCT = nil
+  NO_CHANGE = [].freeze
+
+  def initialize
+    @product_trays = init_product_trays
+    @coin_sorter = CoinSorter.new
+    @amount_deposited = 0
+    @display = nil
+  end
+
+  #
+  # Public Interface
+  #
+  # Consumer Methods
+  def insert_coin(coin)
+    @amount_deposited += @coin_sorter.deposit(coin)
+  rescue CoinSorter::InvalidCoinError => e
+    display_message("Invalid coin: #{e.message}.")
+    @amount_deposited
+  end
+
+  # rubocop: disable Metrics/MethodLength, Metrics/AbcSize
+  def select_product(code)
+    product_tray = @product_trays[code]
+
+    # Validate selection
+    raise InvalidCodeError.new(code) if product_tray.nil?
+
+    # Validate payment
+    overpaid = @amount_deposited - product_tray.price
+    raise PaymentRequiredError.new(overpaid) if overpaid < 0
+
+    # Prepare change and product. Be careful here to rollback any failed actions. That's
+    # why need to make change first.
+    change = @coin_sorter.make_change(overpaid) if overpaid
+    product = product_tray.deliver_product
+
+    # Reset transaction state
+    @amount_deposited = 0
+    clear_display
+
+    # Deliver product and change to user
+    [product, change]
+
+  rescue InvalidCodeError => e
+    display_message("Invalid product code: #{e.message}")
+    [NO_PRODUCT, NO_CHANGE]
+  rescue PaymentRequiredError => e
+    underpaid = -1 * e.message.to_i
+    display_message("Please deposit #{underpaid}p")
+    [NO_PRODUCT, NO_CHANGE]
+  rescue ProductTray::SoldOutError => e
+    display_message("Product sold out: #{e.message}")
+    [NO_PRODUCT, NO_CHANGE]
+  rescue CoinSorter::InsufficientChangeError => e
+    display_message("Insufficient change: #{e.message}")
+    cancel_transaction
+  end
+  # rubocop: enable Metrics/MethodLength, Metrics/AbcSize
+
+  def cancel_transaction
+    coins = @coin_sorter.make_change(@amount_deposited)
+    @amount_deposited = 0
+    [nil, coins]
+  end
+
```

Comment:
This is probably too much:

```
def read_display
  @display
end
```

Or?

```
def read_display
  p @display
end
```

Reply:
Also:

```
def shake_machine_really_hard
  products = []

  @product_trays.each_values do |tray|
    next if tray.empty?
    product << tray.deliver_product if rand > .10
  end
  
  [products, NO_CHANGE]
end
```

Reply:
Couldn't you just use `@display_message = 'Message'` and skip the `display_message` method? You can probably make that an `attr_accessor` instead and save the extra method?

Reply:
I also would not use `puts` / `p` in the vending machine logic itself since that's UI logic which should be separate from the vending machine itself.

Reply:
On second thought, maybe `display_message` is fine if you want to ditch the `attr_accessor` and stick to `attr_reader`. I think `read_display` would be unnecessary since we can just use the `attr_reader`. 

👍 for shaking the machine hard.

Reply:
> I also would not use puts / p in the vending machine logic itself since that's UI logic which should be separate from the vending machine itself.

That's a good point and something I kinda struggled with in doing this. I find it's a common dilemma I face with problems like this: is my code supposed to represent a simulation of the system or is it meant to be more like microcontroller embedded in a system? Ideally, there'd be a lot of overlap between the two in the final solution. My approach here is more simulation.